src/dlsproto/node/neo/request/GetRange.d

```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
+        message, before the connection handler sends the supported code back to
+        the client.
+
+        Note: the initial payload is a slice of the connection's read buffer.
+        This means that when the request-on-conn fiber suspends, the contents of
+        the buffer (hence the slice) may change. It is thus *absolutely
+        essential* that this method does not suspend the fiber. (This precludes
+        all I/O operations on the connection.)
 
         Params:
-            connection = connection to client
-            msg_payload = initial message read from client to begin the request
-                (the request code and version are assumed to be extracted)
+            init_payload = initial message payload read from the connection
 
     ***************************************************************************/
 
-    final public void handle ( RequestOnConn connection, Const!(void)[] msg_payload )
+    public void preSupportedCodeSent ( Const!(void)[] msg_payload )
     {
-        // Send the finished code to the client to indicate end
-        // of the request on the error
-        void sendFinishedCode ()
-        {
-            this.ed.send(
-                ( ed.Payload payload )
-                {
-                    payload.addConstant(MessageType_v1.Finished);
-                }
-            );
-        }
-
-        this.connection = connection;
-        this.ed = connection.event_dispatcher;
-        this.parser = this.ed.message_parser;
-
         cstring channel_name;
         cstring filter_string;
         time_t low, high;
         Filter.FilterMode filter_mode;
 
-        this.ed.send(
-            ( ed.Payload payload )
-            {
-                payload.addConstant(RequestStatusCode.Started);
-            }
-        );
-
         try
         {
-            this.parser.parseBody(msg_payload, channel_name, low, high,
+            this.ed.message_parser.parseBody(msg_payload, channel_name, low, high,
                     filter_string, filter_mode);
 
             if ( !this.prepareChannel(channel_name) ||
                  !this.prepareRange(low, high) ||
                  (filter_string.length > 0 && !this.prepareFilter(filter_mode, filter_string)))
             {
-                sendFinishedCode();
                 return;
             }
 
-            this.value_buffer = this.resources.getVoidBuffer();
-            this.batch_buffer = this.resources.getVoidBuffer();
-            this.compressed_batch = this.resources.getVoidBuffer();
-            this.saved_exception = this.resources.getException();
-            this.lzo = this.resources.getLzo();
-
-            this.run();
+            this.initialised_ok = true;
         }
         catch (Exception e)
         {
             log.error("{}", getMsg(e));
-            sendFinishedCode();
         }
-        finally
+    }
+
+    public void postSupportedCodeSent ()
+    {
+        this.value_buffer = this.resources.getVoidBuffer();
```

Comment:
Small optimization: you could only acquire these resources if the request was initialised ok.